extensions/email/src/main/java/org/mule/extension/email/internal/operations/ListOperation.java

```diff
@@ -0,0 +1,84 @@
+
+import static javax.mail.Folder.READ_ONLY;
+import static org.mule.extension.email.internal.builder.EmailAttributesBuilder.fromMessage;
+import static org.mule.extension.email.internal.util.EmailConnectorUtils.getTextBody;
+import org.mule.extension.email.api.retriever.RetrieverConnection;
+import org.mule.extension.email.api.EmailAttributes;
+import org.mule.extension.email.internal.exception.EmailRetrieverException;
+import org.mule.runtime.api.message.MuleMessage;
+import org.mule.runtime.core.DefaultMuleMessage;
+import org.mule.runtime.core.api.MuleContext;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.function.Predicate;
+
+import javax.mail.Folder;
+import javax.mail.Message;
+import javax.mail.MessagingException;
+
+/**
+ * Represents the list emails operation.
+ *
+ * @since 4.0
+ */
+public final class ListOperation
+{
+
+    /**
+     * Retrieves all the emails in the specified {@code folder}.
+     * <p>
+     * A new {@link MuleMessage} is created for each fetched email from the folder, where
+     * the payload is the text body of the email and the other metadata is carried by
+     * an {@link EmailAttributes} instance.
+     * <p>
+     * For folder implementations (like IMAP) that support fetching without reading the content, if the content
+     * should NOT be read ({@code readContent} = false) the SEEN flag is not going to be setted.
+     *
+     * @param connection  the associated {@link RetrieverConnection}.
+     * @param context     the mule context to create the result message.
+     * @param readContent if should read the email content or not.
+     * @param matcher     a {@link Predicate} of {@link EmailAttributes} used to filter the output list
+     * @return a {@link List} of {@link MuleMessage} carrying all the emails and it's corresponding attributes.
+     */
+    public List<MuleMessage<String, EmailAttributes>> list(RetrieverConnection connection,
+                                                           MuleContext context,
+                                                           boolean readContent,
+                                                           Predicate<EmailAttributes> matcher)
+    {
+        Folder folder = connection.getOpenFolder(READ_ONLY);
+        try
+        {
+            List<MuleMessage<String, EmailAttributes>> list = new LinkedList<>();
+            for (Message m : folder.getMessages())
+            {
+                String body = "";
+                EmailAttributes attributes = fromMessage(m, false);
+                if (matcher.test(attributes))
+                {
+                    if (readContent)
+                    {
+                        body = getTextBody(m);
+                        attributes = fromMessage(m);
+                    }
+                    list.add((MuleMessage) new DefaultMuleMessage(body, null, attributes, context));
+                }
+
+
+            }
+            connection.closeFolder(false);
```

Comment:
doing this should be centralized on the connection provider. Yet another reason to rethink if the connection should be pooled instead of cached, and if the folder should be a parameter instead of state.